AutoResetWrapper integration with gym.make()

balisujohn opened the pull request:
# Description

This is a pull request to integrate the autoreset wrapper into gym.make() with it being disabled by default. The keyword argument is autoreset. If autoreset is set to False or a value is not provided for auto_reset in gym.make, then the env will not be wrapped with the autoreset wrapper. 

Some things to note that might be relevant for the code review are current the only test is for cartpole is because it is guaranteed to terminate in a relative short number of time-steps. I'm open to suggestions for how to improve test coverage.

Fixes https://github.com/openai/gym/issues/2564

## Type of change

Please delete options that are not relevant.

- [ ] This change requires a documentation update

# Checklist:

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

jkterry1 commented:
Two comments:
-Autoreset should default to off, is that not done here? I'm not seeing it in the code, perhaps I'm going crazy
-This needs to be added to the docs

balisujohn commented:
The env spec class has a field autoreset which defaults to false, this is only changed if the autoreset argument is supplied to the make function.

Where in the docs do you want the integration mentioned, aside from a entry being added to the wrappers page for AutoResetwrapper? Also does it say somewhere how to build the docs locally?

balisujohn commented:
Aight added a documentation pull https://github.com/Farama-Foundation/gym-docs/pull/108

RedTachyon reviewed (commented):
While reviewing this I realized there might be a more fundamental problem with how kwargs are passed into `gym.make` to apply wrappers, so could you please do the same thing as in L137-L140 to `max_episode_steps` so that this also works as a keyword argument to `make`? This pattern might be changed in the future, but for now that's the only way to make these wrappers actually work.

Otherwise, it's straightforward enough, so my main comments are about tests.

balisujohn commented on tests/wrappers/test_autoreset.py:
```diff
@@ -62,6 +64,46 @@ def test_autoreset_reset_info():
+    obs = env.reset(seed=0)
+    env.action_space.seed(0)
+
+    env.env.reset = MagicMock(side_effect=env.env.reset)
```
That's correct regarding how MagicMock works, and I updated the code to use unwrapped.

balisujohn commented on tests/wrappers/test_autoreset.py:
```diff
@@ -62,6 +64,46 @@ def test_autoreset_reset_info():
     assert isinstance(info, dict)
 
 
+def test_make_autoreset():
```
My initial worry was that some environments might not be guaranteed to terminate with random actions, but actually I think all environments are, since all except blackjack-v1 have max_episode_steps specified, and blackjack will terminate in less than some finite number of steps no matter what actions are taken. So I implemented the pattern so now it's applied to all first-party environments. It adds something like 30 seconds to the test battery.

balisujohn commented on tests/wrappers/test_autoreset.py:
```diff
@@ -62,6 +64,46 @@ def test_autoreset_reset_info():
+
+    assert env.env.reset.called
+
+    env = gym.make("CartPole-v1", autoreset=False)
```
done

balisujohn commented on gym/envs/registration.py:
```diff
@@ -142,15 +148,23 @@ def make(self, **kwargs) -> Env:
 
             env = TimeLimit(env, max_episode_steps=env.spec.max_episode_steps)
+
+        if self.autoreset == True:
```
done

balisujohn commented:
I addressed the code review, except I'd prefer not to make any changes to the max_episode_steps field in this PR, since I don't fully understand the registration system. It seems like max_episode_steps is supplied in __init__.py in the gym/envs folder, so it's not clear to me that it's supposed to be possible to specify that as a direct argument to make(not that I'm ideological about whether it should be possible). Otherwise, the pre-commit is failing, but the failure is coming from files I didn't change, the pre-commit seems to be broken right now, since the last few commits into master did not pass the pre-commit.

balisujohn commented:
alright looks like the pre-commit is fixed with the new merged pull request.

pseudo-rnd-thoughts commented:
@balisujohn @jkterry1 I just checked my build action that was failing (#2727) I saw that there are python segmentation faults happening that are not getting thrown as error. 
I have had a quick look through previous PR and can't see the seg faults there but I can't confirm that this is the problem. 
Running the master locally, I dont have this issue but @jkterry1 do we want to revert to previous master while this issue is not addressed yet?
